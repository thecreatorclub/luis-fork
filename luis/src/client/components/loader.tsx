import * as React from 'react';

import { ProxyStore } from '../models/proxy_store';

type Props = {
  component: React.ComponentType;
  proxyConfig: any;
};

export class Loader extends React.Component<Props> {
  static displayName = 'Loader';

  // componentWillReceiveProps({ proxies, component }: Props) {
  //   if (this.props.component !== component || proxies !== this.props.proxies) {
  //     this.firstProxy = createProxyLinkedList([...proxies, this.props.component]);
  //   }
  // }

  render() {
    // const { firstProxy } = this;
    const { component, proxyConfig } = this.props;
    const proxies = ProxyStore.findProxies(proxyConfig);

    const firstProxy: any = createProxyLinkedList([...proxies, this.props.component]);

    return (
      <firstProxy.value
        proxyConfig={proxyConfig}
        nextProxy={firstProxy.next()}
        fixture={component}
        onComponentRef={noOp}
        onFixtureUpdate={noOp}
      />
    );
  }
}

type LinkedItem<Item> = {
  value: Item;
  next: () => LinkedItem<Item>;
};

function createProxyLinkedList<T>(items: T[]) {
  function advanceList(toIndex: number): LinkedItem<T> {
    return {
      value: items[toIndex],
      next: function getNextItem(nextIndex: number): LinkedItem<T> {
        return advanceList(nextIndex);
      }.bind(null, toIndex + 1)
    };
  }

  return advanceList(0);
}

function noOp() {}
