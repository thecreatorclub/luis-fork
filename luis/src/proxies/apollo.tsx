// Shamefully taken from Cosmos and adapted to our needs:
// https://github.com/react-cosmos/react-cosmos

import * as React from 'react';

import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { HttpLink } from 'apollo-link-http';
import { ApolloProvider } from 'react-apollo';
import { createFixtureLink } from './helpers/fixture_link';
import { ProxyProps } from './helpers/proxy_types';

const defaults = {
  fixtureKey: 'apollo'
};

type ApolloOptions = {
  endpoint?: string;
  clientOptions?: any;
};

export function createApolloProxy(options: ApolloOptions) {
  global.process = global.process || ({ env: { NODE_ENV: 'test' } } as any);

  const { endpoint, clientOptions } = {
    ...defaults,
    ...options
  };

  class ApolloProxy extends React.Component<ProxyProps> {
    client: ApolloClient<any>;

    constructor(props: ProxyProps) {
      super(props);

      if (!endpoint && !clientOptions) {
        throw new Error(
          `
It looks like the Apollo Proxy is not configured!
Give it:
- a GraphQL endpoint to send GraphQL operations to;
- a Apollo Client configuration object (maybe the one you use in your app?);
Read more at: https://github.com/react-cosmos/react-cosmos#react-apollo-graphql.`
        );
      }

      const apolloFixture = this.props.proxyConfig.apollo || {};
      const mockKeys = ['resolveWith', 'failWith'];

      const fixtureApolloKeys = flatObjectKeys(mockKeys, apolloFixture);

      const isMockedFixture = Boolean(mockKeys.find(key => fixtureApolloKeys.includes(key)));

      let options = clientOptions || {
        cache: new InMemoryCache()
      };

      if (isMockedFixture) {
        const fixtureCache = new InMemoryCache(options.cache.config);

        options = {
          ...options,
          // ensure that the cache is not persisted between mocked fixtures
          cache: fixtureCache,
          link: createFixtureLink({
            apolloFixture,
            cache: fixtureCache,
            fixture: this.props.fixture
          })
        };
      } else {
        options = {
          ...options,

          link: new HttpLink({ uri: endpoint })
        };
      }

      this.client = new ApolloClient(options);

      // enable the Apollo Client DevTools to recognize the Apollo Client instance
      if (typeof window !== 'undefined') {
        (window as any).__APOLLO_CLIENT__ = this.client;
      }
    }

    render() {
      const { value: NextProxy, next } = this.props.nextProxy;

      return (
        <ApolloProvider client={this.client}>
          <NextProxy {...this.props} nextProxy={next()} />
        </ApolloProvider>
      );
    }
  }

  return {
    proxy: ApolloProxy as any,
    key: 'apollo'
  };
}

// Utility to find mock keys inside a fixture
function flatObjectKeys(keys: string[], object: any, digNestedObjects = true): string[] {
  return Object.keys(object).reduce((list, key) => {
    const value = object[key];
    if (!keys.includes(key) && typeof value === 'object' && digNestedObjects) {
      // only "dig" one level deep
      return [...list, ...flatObjectKeys(keys, value, false)];
    }

    return [...list, key];
  }, []);
}
