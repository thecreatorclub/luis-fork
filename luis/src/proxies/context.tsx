// Shamefully taken from Cosmos and adapted to our needs:
// https://github.com/react-cosmos/react-cosmos/blob/master/packages/react-cosmos-context-proxy/src/index.js

// @flow

import * as React from 'react';
import { ProxyProps } from './helpers/proxy_types';

type Options = {
  childContextTypes: {
    [index: string]: any; // React doesn't really expose types for prop-types
  };
};

export function createContextProxy({ childContextTypes }: Options) {
  class ContextProxy extends React.Component<ProxyProps> {
    static childContextTypes = childContextTypes;

    getChildContext() {
      return this.props.proxyConfig && this.props.proxyConfig.context
        ? this.props.proxyConfig.context
        : {};
    }

    render() {
      const { nextProxy, ...rest } = this.props;
      const { value: NextProxy, next } = nextProxy;

      return <NextProxy {...rest} nextProxy={next()} />;
    }
  }

  return {
    proxy: ContextProxy as any,
    key: 'context'
  };
}
