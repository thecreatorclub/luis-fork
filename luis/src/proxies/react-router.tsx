// Shamefully taken from Cosmos and adapted to our needs:
// https://github.com/react-cosmos/react-cosmos

import React from 'react';
import { MemoryRouter, Route } from 'react-router';
import LocationInterceptor from './helpers/location_interceptor';
import urlParser from 'url';
import { ProxyProps } from './helpers/proxy_types';

export function createRouterProxy() {
  const RouterProxy = (props: ProxyProps) => {
    const { nextProxy, onFixtureUpdate } = props;
    const { value: NextProxy, next } = nextProxy as any;
    const { route, url, locationState, provideRouterProps } = props.proxyConfig;
    const nextProxyEl = <NextProxy {...props} nextProxy={next()} />;

    if (locationState && !url) {
      throw new Error('Must provide a url in fixture to use locationState');
    }

    if (!url) {
      return nextProxyEl;
    }

    const handleUrlChange = (url: string) => {
      onFixtureUpdate({ url });
    };

    const handleLocationStateChange = (locationState: any) => {
      onFixtureUpdate({ locationState });
    };

    const location = buildLocation(url, locationState);

    return (
      // keyLength=0 generates empty "" history keys, leading to deterministic
      // test snapshots https://github.com/react-cosmos/react-cosmos/pull/803
      <MemoryRouter keyLength={0} initialEntries={[location]}>
        <LocationInterceptor
          onUrlChange={handleUrlChange}
          onLocationStateChange={handleLocationStateChange}
        >
          {route ? (
            <Route
              path={route}
              render={(routerProps: any) => {
                if (provideRouterProps) {
                  const newProxyProps = {
                    ...props,
                    fixture: {
                      ...props.fixture,
                      props: { ...(props.fixture as any).props, ...routerProps }
                    }
                  };

                  return <NextProxy {...newProxyProps} nextProxy={next()} />;
                }

                return nextProxyEl;
              }}
            />
          ) : (
            nextProxyEl
          )}
        </LocationInterceptor>
      </MemoryRouter>
    );
  };

  return RouterProxy;
}

function buildLocation(url: string, locationState: any) {
  const { pathname, search, hash } = urlParser.parse(url);

  return {
    pathname,
    search,
    hash,
    key: 'mocked',
    state: locationState
  };
}
