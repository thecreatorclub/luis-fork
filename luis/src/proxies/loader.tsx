import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Loader } from '../client/components/loader';

export function load(config: { component: React.ComponentType; [index: string]: any }) {
  const { component, ...proxyConfig } = config;
  return <Loader component={component} proxyConfig={proxyConfig} />;
}

export function render(config: { component: React.ComponentType; [index: string]: any }) {
  return renderer.create(load(config));
}

export async function renderApollo(
  config: { component: React.ComponentType; [index: string]: any },
  waitLoad = true
) {
  const result = renderer.create(load(config));
  if (waitLoad) {
    await wait();
  }
  return result;
}

export function wait() {
  return new Promise(r => setTimeout(r, 0));
}
