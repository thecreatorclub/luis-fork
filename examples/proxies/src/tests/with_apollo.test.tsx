import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { load, wait, renderApollo } from 'luis/proxies/loader';

describe('With Apollo', function() {
  const Component: React.FC = () => {
    return (
      <Query
        query={gql`
          query People {
            people
          }
        `}
      >
        {props => {
          if (props.loading) {
            return <div>Loading ...</div>;
          }
          if (props.error) {
            return <pre>Error: {JSON.stringify(props.error, null, 2)}</pre>;
          }
          return (
            <ul>
              <li>Fail</li>
              {props.data.people.map((p: string, i: number) => (
                <li key={i}>{p}</li>
              ))}
            </ul>
          );
        }}
      </Query>
    );
  };

  const config = {
    component: Component,
    apollo: {
      // mocked response for the query named PostsForAuthor
      People: {
        resolveWith: {
          people: ['Tomas', 'Valeria']
        }
      }
    }
  };

  it('renders', async function() {
    const render = renderer.create(load(config));
    expect(render).toMatchSnapshot('Loading');

    await wait();

    expect(render).toMatchSnapshot('Rendered');
  });

  it('renders with helper', async function() {
    // just a helper method that renders apollo container
    // skips the loading phase by calling 'wait'
    const render = await renderApollo(config);
    expect(render).toMatchSnapshot('Rendered');
  });

  it('renders with error', async function() {
    // just a helper method that renders apollo container
    // skips the loading phase by calling 'wait'
    const render = await renderApollo({
      component: Component,
      apollo: {
        // mocked response for the query named PostsForAuthor
        People: {
          resolveWith: {
            errors: [{ message: ['This is my error message'] }]
          }
        }
      }
    });
    expect(render).toMatchSnapshot('Error');
  });

  addStory(config);
});
