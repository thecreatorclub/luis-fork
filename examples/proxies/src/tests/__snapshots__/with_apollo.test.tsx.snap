// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Apollo renders 1`] = `
<div>
  Loading ...
</div>
`;

exports[`With Apollo renders with error: Error 1`] = `
<pre>
  {
  "graphQLErrors": [],
  "networkError": {},
  "message": "Network error: Missing mock for operation 'People'. You mocked only: ['failWith']"
}
</pre>
`;

exports[`With Apollo renders with error: Rendered 1`] = `
<pre>
  {
  "graphQLErrors": [
    {
      "path": [
        "author"
      ],
      "message": [
        "Author id -1 not found"
      ],
      "locations": [
        {
          "line": 1,
          "column": 0
        }
      ]
    }
  ],
  "networkError": null,
  "message": "GraphQL error: Author id -1 not found"
}
</pre>
`;

exports[`With Apollo renders with helper: Rendered 1`] = `
<ul>
  <li>
    Fail
  </li>
  <li>
    Tomas
  </li>
  <li>
    Valeria
  </li>
</ul>
`;

exports[`With Apollo renders: Loading 1`] = `
<div>
  Loading ...
</div>
`;

exports[`With Apollo renders: Rendered 1`] = `
<ul>
  <li>
    Fail
  </li>
  <li>
    Tomas
  </li>
  <li>
    Valeria
  </li>
</ul>
`;
